mempool/src/shared_mempool/network.rs

```diff
@@ -318,149 +335,190 @@ impl MempoolNetworkInterface {
-                    metric_label = if Some(id) == expired {
+                    let metric_label = if Some(id) == expired_batch_id {
                         Some(counters::EXPIRED_BROADCAST_LABEL)
                     } else {
                         Some(counters::RETRY_BROADCAST_LABEL)
                     };
 
                     let txns = mempool.timeline_range(id.0, id.1);
-                    (*id, txns)
+                    (*id, txns, metric_label)
                 }
                 None => {
                     // Fresh broadcast
                     let (txns, new_timeline_id) = mempool.read_timeline(
                         state.timeline_id,
                         self.mempool_config.shared_mempool_batch_size,
                     );
-                    (BatchId(state.timeline_id, new_timeline_id), txns)
+                    (BatchId(state.timeline_id, new_timeline_id), txns, None)
                 }
             };
 
-            batch_id = new_batch_id;
-            transactions = new_transactions;
-        }
-
         if transactions.is_empty() {
-            return;
+            return Err(BroadcastError::NoTransactions(peer));
         }
 
-        let network_sender = smp.network_interface.sender();
+        Ok((batch_id, transactions, metric_label))
+    }
 
-        let num_txns = transactions.len();
-        if let Err(e) = network_sender.send_to(
-            peer,
-            MempoolSyncMsg::BroadcastTransactionsRequest {
-                request_id: bcs::to_bytes(&batch_id).expect("failed BCS serialization of batch ID"),
-                transactions,
-            },
-        ) {
+    /// Sends a batch to the given `Peer`
+    async fn send_batch(
+        &self,
+        peer: PeerNetworkId,
+        batch_id: BatchId,
+        transactions: Vec<SignedTransaction>,
+    ) -> Result<(), BroadcastError> {
+        let request = MempoolSyncMsg::BroadcastTransactionsRequest {
+            request_id: bcs::to_bytes(&batch_id).expect("failed BCS serialization of batch ID"),
+            transactions,
+        };
+
+        if let Err(e) = self.sender.send_to(peer, request) {
             counters::network_send_fail_inc(counters::BROADCAST_TXNS);
-            error!(
-                LogSchema::event_log(LogEntry::BroadcastTransaction, LogEvent::NetworkSendFail)
-                    .peer(&peer)
-                    .error(&e.into())
-            );
-            return;
+            return Err(BroadcastError::NetworkError(peer, e.into()));
         }
+        Ok(())
+    }
+
+    /// Updates the local tracker for a broadcast.  This is used to handle `DirectSend` tracking of
+    /// responses
+    fn update_broadcast_state(
+        &self,
+        peer: PeerNetworkId,
+        batch_id: BatchId,
+        send_time: SystemTime,
+    ) -> Result<usize, BroadcastError> {
+        let mut sync_states = self.sync_states.write_lock();
+        let state = if let Some(state) = sync_states.get_mut(&peer) {
```

Comment:
nit: could you not use something like `ok_or` or `ok_or_else` to unwrap this?, e.g., something like:

```
let state = sync_states.get_mut(&peer).ok_or(BroadcastError::PeerNotFound(peer))?;
```

Reply:
Yeah, this originally didn't have errors, which made the flow not able to do that .  I can change it.